// veldt/scripts/render/color/ColorRamp.js
'use strict';

const map = require('lodash/map');
const sum = require('lodash/sum');

const NUM_GRADIENT_STEPS = 256;

function rgb2lab(rgb) {
	const r = rgb[0] > 0.04045 ? Math.pow((rgb[0] + 0.055) / 1.055, 2.4) : rgb[0] / 12.92;
	const g = rgb[1] > 0.04045 ? Math.pow((rgb[1] + 0.055) / 1.055, 2.4) : rgb[1] / 12.92;
	const b = rgb[2] > 0.04045 ? Math.pow((rgb[2] + 0.055) / 1.055, 2.4) : rgb[2] / 12.92;
	// Observer. = 2°, Illuminant = D65
	let x = r * 0.4124564 + g * 0.3575761 + b * 0.1804375;
	let y = r * 0.2126729 + g * 0.7151522 + b * 0.0721750;
	let z = r * 0.0193339 + g * 0.1191920 + b * 0.9503041;
	x = x / 0.95047; // Observer= 2°, Illuminant= D65
	y = y / 1.00000;
	z = z / 1.08883;
	x = x > 0.008856 ? Math.pow(x, 1 / 3) : (7.787037 * x) + (16 / 116);
	y = y > 0.008856 ? Math.pow(y, 1 / 3) : (7.787037 * y) + (16 / 116);
	z = z > 0.008856 ? Math.pow(z, 1 / 3) : (7.787037 * z) + (16 / 116);
	return [
		(116 * y) - 16,
		500 * (x - y),
		200 * (y - z),
		rgb[3]
	];
}

function lab2rgb(lab) {
	let y = (lab[0] + 16) / 116;
	let x = y + lab[1] / 500;
	let z = y - lab[2] / 200;
	x = x > 0.206893034 ? x * x * x : (x - 4 / 29) / 7.787037;
	y = y > 0.206893034 ? y * y * y : (y - 4 / 29) / 7.787037;
	z = z > 0.206893034 ? z * z * z : (z - 4 / 29) / 7.787037;
	x = x * 0.95047; // Observer= 2°, Illuminant= D65
	y = y * 1.00000;
	z = z * 1.08883;
	let r = x * 3.2404542 + y * -1.5371385 + z * -0.4985314;
	let g = x * -0.9692660 + y * 1.8760108 + z * 0.0415560;
	let b = x * 0.0556434 + y * -0.2040259 + z * 1.0572252;
	r = r > 0.00304 ? 1.055 * Math.pow(r, 1 / 2.4) - 0.055 : 12.92 * r;
	g = g > 0.00304 ? 1.055 * Math.pow(g, 1 / 2.4) - 0.055 : 12.92 * g;
	b = b > 0.00304 ? 1.055 * Math.pow(b, 1 / 2.4) - 0.055 : 12.92 * b;
	return [Math.max(Math.min(r, 1), 0), Math.max(Math.min(g, 1), 0), Math.max(Math.min(b, 1), 0), lab[3]];
}

function distance(c1, c2) {
	return Math.sqrt(
		(c1[0] - c2[0]) * (c1[0] - c2[0]) +
		(c1[1] - c2[1]) * (c1[1] - c2[1]) +
		(c1[2] - c2[2]) * (c1[2] - c2[2]) +
		(c1[3] - c2[3]) * (c1[3] - c2[3]));
}

const buildFlatLookupTable = function(color) {
	const output = [];
	output.push(color[0]);
	output.push(color[1]);
	output.push(color[2]);
	output.push(color[3]);
	return output;
};

// Interpolate between a set of colors using even perceptual distance and interpolation in CIE L*a*b* space
const buildPerceptualLookupTable = function(baseColors) {
	const outputGradient = new Uint8Array(NUM_GRADIENT_STEPS*4);
	// Calculate perceptual spread in L*a*b* space
	const labs = map(baseColors, color => {
		return rgb2lab([color[0] / 255, color[1] / 255, color[2] / 255, color[3] / 255]);
	});
	let distances = map(labs, (color, index, colors) => {
		return index > 0 ? distance(color, colors[index - 1]) : 0;
	});
	// Calculate cumulative distances in [0,1]
	const totalDistance = sum(distances);
	distances = map(distances, d => {
		return d / totalDistance;
	});
	let distanceTraversed = 0;
	let key = 0;
	let progress;
	let stepProgress;
	let rgb;
	for (let i = 0; i < NUM_GRADIENT_STEPS; i++) {
		progress = i / (NUM_GRADIENT_STEPS - 1);
		if (progress > distanceTraversed + distances[key + 1] && key + 1 < labs.length - 1) {
			key += 1;
			distanceTraversed += distances[key];
		}
		stepProgress = (progress - distanceTraversed) / distances[key + 1];
		rgb = lab2rgb([
			labs[key][0] + (labs[key + 1][0] - labs[key][0]) * stepProgress,
			labs[key][1] + (labs[key + 1][1] - labs[key][1]) * stepProgress,
			labs[key][2] + (labs[key + 1][2] - labs[key][2]) * stepProgress,
			labs[key][3] + (labs[key + 1][3] - labs[key][3]) * stepProgress
		]);
		outputGradient[i * 4] = rgb[0] * 255;
		outputGradient[i * 4 + 1] = rgb[1] * 255;
		outputGradient[i * 4 + 2] = rgb[2] * 255;
		outputGradient[i * 4 + 3] = rgb[3] * 255;
	}
	return outputGradient;
};

const COOL = buildPerceptualLookupTable([
	[0x04, 0x20, 0x40, 0x50],
	[0x08, 0x40, 0x81, 0x7f],
	[0x08, 0x68, 0xac, 0xff],
	[0x2b, 0x8c, 0xbe, 0xff],
	[0x4e, 0xb3, 0xd3, 0xff],
	[0x7b, 0xcc, 0xc4, 0xff],
	[0xa8, 0xdd, 0xb5, 0xff],
	[0xcc, 0xeb, 0xc5, 0xff],
	[0xe0, 0xf3, 0xdb, 0xff],
	[0xf7, 0xfc, 0xf0, 0xff]
]);

const HOT = buildPerceptualLookupTable([
	[0x40, 0x00, 0x13, 0x50],
	[0x80, 0x00, 0x26, 0x7f],
	[0xbd, 0x00, 0x26, 0xff],
	[0xe3, 0x1a, 0x1c, 0xff],
	[0xfc, 0x4e, 0x2a, 0xff],
	[0xfd, 0x8d, 0x3c, 0xff],
	[0xfe, 0xb2, 0x4c, 0xff],
	[0xfe, 0xd9, 0x76, 0xff],
	[0xff, 0xed, 0xa0, 0xff]
]);

const VERDANT = buildPerceptualLookupTable([
	[0x00, 0x40, 0x26, 0x50],
	[0x00, 0x5a, 0x32, 0x7f],
	[0x23, 0x84, 0x43, 0xff],
	[0x41, 0xab, 0x5d, 0xff],
	[0x78, 0xc6, 0x79, 0xff],
	[0xad, 0xdd, 0x8e, 0xff],
	[0xd9, 0xf0, 0xa3, 0xff],
	[0xf7, 0xfc, 0xb9, 0xff],
	[0xff, 0xff, 0xe5, 0xff]
]);

const SPECTRAL = buildPerceptualLookupTable([
	[0x26, 0x1a, 0x40, 0x50],
	[0x44, 0x2f, 0x72, 0x7f],
	[0xe1, 0x2b, 0x02, 0xff],
	[0x02, 0xdc, 0x01, 0xff],
	[0xff, 0xd2, 0x02, 0xff],
	[0xff, 0xff, 0xff, 0xff]
]);

const TEMPERATURE = buildPerceptualLookupTable([
	[0x00, 0x16, 0x40, 0x50],
	[0x00, 0x39, 0x66, 0x7f],
	[0x31, 0x3d, 0x66, 0xff],
	[0xe1, 0x2b, 0x02, 0xff],
	[0xff, 0xd2, 0x02, 0xff],
	[0xff, 0xff, 0xff, 0xff]
]);

const GREYSCALE = buildPerceptualLookupTable([
	[0x00, 0x00, 0x00, 0x7f],
	[0x40, 0x40, 0x40, 0xff],
	[0xff, 0xff, 0xff, 0xff]
]);

const GOLD = buildPerceptualLookupTable([
	[ 0x84, 0x54, 0x0F, 0xFF ],
	[ 0xA6, 0x7B, 0x3E, 0xFF ],
	[ 0xC9, 0xA3, 0x6D, 0xFF ],
	[ 0xEC, 0xCB, 0x9C, 0xFF ],
	[ 0xff, 0xff, 0xff, 0xff ]
]);

const FLAT = buildFlatLookupTable([0xff, 0xff, 0xff, 0xff]);

const buildLookupFunction = function(RAMP) {
	return function(scaledValue, inColor) {
		const index = Math.floor(scaledValue * (RAMP.length / 4 - 1));
		inColor[0] = RAMP[index * 4];
		inColor[1] = RAMP[index * 4 + 1];
		inColor[2] = RAMP[index * 4 + 2];
		inColor[3] = RAMP[index * 4 + 3];
		return inColor;
	};
};

const colorTables = {
	cool: COOL,
	hot: HOT,
	verdant: VERDANT,
	spectral: SPECTRAL,
	temperature: TEMPERATURE,
	gold: GOLD,
	grey: GREYSCALE,
	flat: FLAT
};

const colorRamps = {
	cool: buildLookupFunction(COOL),
	hot: buildLookupFunction(HOT),
	verdant: buildLookupFunction(VERDANT),
	spectral: buildLookupFunction(SPECTRAL),
	temperature: buildLookupFunction(TEMPERATURE),
	gold: buildLookupFunction(GOLD),
	grey: buildLookupFunction(GREYSCALE),
	flat: buildLookupFunction(FLAT)
};

const getTable = function(type) {
	if (!colorTables[type]) {
		throw `Color table ${type} does not exist`;
	}
	return colorTables[type];
};

const getFunc = function(type) {
	if (!colorRamps[type]) {
		throw `Color table ${type} does not exist`;
	}
	return colorRamps[type];
};

const createRamp = function(type, baseColors) {
	colorTables[type] = buildPerceptualLookupTable(baseColors);
	colorRamps[type] = buildLookupFunction(colorTables[type]);
	return colorRamps[type];
};

const getBuckets = function(type, numBuckets) {
	const ramp = getFunc(type);
	const buckets = [];
	const color = [];
	for (let i = 0; i<numBuckets; i++) {
		ramp(i/(numBuckets-1), color);
		buckets.push([
			color[0] / 255,
			color[1] / 255,
			color[2] / 255,
			color[3] / 255
		]);
	}
	return buckets;
};

module.exports = {
	getTable: getTable,
	getFunc: getFunc,
	getBuckets: getBuckets,
	createRamp: createRamp,
	NUM_GRADIENT_STEPS: NUM_GRADIENT_STEPS
};
